#### Project - 4
#### Travelling Salesman problem
Prepared for: ExxpertSCM Thoughtware Training Pvt Ltd.


Guided by:  Mr. Pattabhi Raman


The travelling salesman problem (also called the travelling salesperson problem or TSP) is a classic optimization problem that seeks to find the shortest possible route for a salesman to visit a set of cities and return to the starting point, it asks the following question: "Given a list of cities and the distances between each pair of cities, what is the shortest possible route that visits each city exactly once and returns to the origin city?" It can be solved using different methods, here we consider using TSP library in python.

Dataset :
We arev given with a disatnce matrix as our dataset along with the name of the cities to be travelled. we consider the Bengaluru  as our starting point and the final destination should be the origin city itself.In between the there are 13 other cities whose distance with each other is given in KMs.
the cities we consider are:
Bengaluru
Chennai
Kochi
Trivandrum
Salem
Pondicherry
Madurai
Mysore
Coimbatore
Mangaluru
Vellore
Kozhikode
Shivamogga
Hubballi


Constraints:
In this optimization problem we are given with certain constraints they are, the starting and returning point should be the warehouse. Each city should be travelled only once. Also only one set of route is needed.


In [ ]:
pip install python-tsp

In [ ]:
import pandas as pd
import numpy as np
from python_tsp.exact import solve_tsp_dynamic_programming
import io
from google.colab import files
uploaded = files.upload()

Saving TSP_dataset.xlsx to TSP_dataset (1).xlsx


In [ ]:
## reading dataset
data = pd.read_excel(io.BytesIO(uploaded['TSP_dataset (1).xlsx']))
data

location   Bengaluru  Chennai  Kochi  Trivandrum  Salem  Pondicherry  \
0     Bengaluru        0.0    345.7  547.4       735.9  201.1        313.2   
1       Chennai      345.7      0.0  690.1       778.5  344.7        150.8   
2         Kochi      547.4    690.1    0.0       199.3  349.2        568.6   
3    Trivandrum      735.9    778.5  199.3         0.0  537.0        657.9   
4         Salem      201.1    344.7  349.2       537.0    0.0        222.4   
5   Pondicherry      313.2    150.8  568.6       657.9  222.4          0.0   
6       Madurai      433.5    455.6  265.1       318.5  233.5        332.3   
7        Mysore      143.2    480.9  394.1       680.9  335.4        446.0   
8    Coimbatore      362.9    505.7  189.3       387.4  167.0        382.3   
9     Mangaluru      350.7    704.0  418.2       610.0  558.5        669.1   
10      Vellore      211.3    137.7  575.6       766.3  229.3        153.6   
11    Kozhikode      355.9    683.9  182.4       374.2  345.3        560.5   
12   Shivamogga      306.4    659.7  642.0      1051.2  514.2        624.8   
13     Hubballi      412.1    765.4  861.7      1156.9  619.9        730.5   

    Madurai  Mysore  Coimbatore  Mangaluru  Vellore  Kozhikode  Shivamogga  \
0     433.5   143.2       362.9      350.7    211.3      355.9       306.4   
1     455.6   480.9       505.7      704.0    137.7      683.9       659.7   
2     265.1   394.1       189.3      418.2    575.6      182.4       642.0   
3     318.5   680.9       387.4      610.0    766.3      374.2      1051.2   
4     233.5   335.4       167.0      558.5    229.3      345.3       514.2   
5     332.3   446.0       382.3      669.1    153.6      560.5       624.8   
6       0.0   378.1       211.4      636.3    400.7      359.7       748.4   
7     378.1     0.0       198.8      255.2    352.6      209.6       243.5   
8     211.4   198.8         0.0      457.0    391.5      183.2       446.8   
9     636.3   255.2       457.0        0.0    560.4      232.6       200.8   
10    400.7   352.6       391.5      560.4      0.0      558.6       524.7   
11    359.7   209.6       183.2      232.6    558.6        0.0       403.7   
12    748.4   243.5       446.8      200.8    524.7      403.7         0.0   
13    854.1   496.5       666.4      361.8    630.4      599.9       210.7   

    Hubballi  
0      412.1  
1      765.4  
2      861.7  
3     1156.9  
4      619.9  
5      730.5  
6      854.1  
7      496.5  
8      666.4  
9      361.8  
10     630.4  
11     599.9  
12     210.7  
13       0.0

In [ ]:
data.shape

(14, 15)

In [ ]:
data.columns

Index(['location ', 'Bengaluru', 'Chennai', 'Kochi', 'Trivandrum', 'Salem',
       'Pondicherry', 'Madurai', 'Mysore', 'Coimbatore', 'Mangaluru',
       'Vellore', 'Kozhikode', 'Shivamogga', 'Hubballi'],
      dtype='object')

In [ ]:
df = data.set_index('location', inplace=True)


KeyError: ignored

In [ ]:
distance_matrix = df.values
distance_matrix

array([['Bengaluru', 0.0, 345.7, 547.4, 735.9, 201.1, 313.2, 433.5,
        143.2, 362.9, 350.7, 211.3, 355.9, 306.4, 412.1],
       ['Chennai', 345.7, 0.0, 690.1, 778.5, 344.7, 150.8, 455.6, 480.9,
        505.7, 704.0, 137.7, 683.9, 659.7, 765.4],
       ['Kochi', 547.4, 690.1, 0.0, 199.3, 349.2, 568.6, 265.1, 394.1,
        189.3, 418.2, 575.6, 182.4, 642.0, 861.7],
       ['Trivandrum', 735.9, 778.5, 199.3, 0.0, 537.0, 657.9, 318.5,
        680.9, 387.4, 610.0, 766.3, 374.2, 1051.2, 1156.9],
       ['Salem', 201.1, 344.7, 349.2, 537.0, 0.0, 222.4, 233.5, 335.4,
        167.0, 558.5, 229.3, 345.3, 514.2, 619.9],
       ['Pondicherry', 313.2, 150.8, 568.6, 657.9, 222.4, 0.0, 332.3,
        446.0, 382.3, 669.1, 153.6, 560.5, 624.8, 730.5],
       ['Madurai', 433.5, 455.6, 265.1, 318.5, 233.5, 332.3, 0.0, 378.1,
        211.4, 636.3, 400.7, 359.7, 748.4, 854.1],
       ['Mysore', 143.2, 480.9, 394.1, 680.9, 335.4, 446.0, 378.1, 0.0,
        198.8, 255.2, 352.6, 209.6, 243.5, 496.5],
  

In [ ]:
optimal_route, distance = solve_tsp_dynamic_programming(distance_matrix)

TypeError: ignored

In [ ]:
print(optimal_route)

[0, 7, 12, 13, 9, 11, 2, 3, 6, 8, 4, 5, 1, 10]


In [ ]:
print("The minimum distance taken for covering all cities: {:.1f} KM".format(distance))

The minimum distance taken for covering all cities: 2992.6 KM


In [ ]:
Locations = df.index.tolist()

In [ ]:
from geopy.geocoders import Nominatim
import folium

# Assuming you have a list called 'locations' containing the indices of the cities
# Assuming you have the optimal route order stored in a list called 'optimal_route'
# Assuming you have a DataFrame called 'df' containing the distance matrix and city names

# Create a map
m = folium.Map(location=[0, 0], zoom_start=2)

# Create a geocoder object
geolocator = Nominatim(user_agent="my_app")

# Plot the optimal route on the map
for i in range(len(optimal_route) - 1):
    index1 = optimal_route[i]
    index2 = optimal_route[i+1]

    # Get the city names from the DataFrame using the indices
    city1 = df.loc[locations[index1], 'location']
    city2 = df.loc[locations[index2], 'location']

    # Get the coordinates of the cities
    location1 = geolocator.geocode(city1)
    location2 = geolocator.geocode(city2)

    # Check if coordinates are found
    if location1 is not None and location2 is not None:
        # Plot a line between the cities
        folium.PolyLine(locations=[(location1.latitude, location1.longitude), (location2.latitude, location2.longitude)], color='red').add_to(m)

# Save the map as an HTML file
m.save('optimal_route_map.html')

# Display the map
m


KeyError: ignored

The ouput of the TSP gives us the order in which the saleman need to visit the cities.

The order in which the salesman need to travel is : 0->7->12->13->9->11->2->3->6->8->4->5->1->10->0.  

The order of the city: Banglore-> Mysore-> Shivamogga-> Hubballi-> Mangluru-> Kozhikode-> Kochi-> Trivandrum-> Madurai-> Coimbatore-> Salem-> Pondicherry-> Chennai-> Vellore-> Banglore.


The distance that the salesman need to travel is 2992.6 km to cover all the cities and reach back the origin city.